lib/services/local/saml_idp_service_provider.go

```diff
@@ -94,9 +94,41 @@ func (s *SAMLIdPServiceProviderService) GetSAMLIdPServiceProvider(ctx context.Co
 
 // CreateSAMLIdPServiceProvider creates a new SAML IdP service provider resource.
 func (s *SAMLIdPServiceProviderService) CreateSAMLIdPServiceProvider(ctx context.Context, sp types.SAMLIdPServiceProvider) error {
+	// Check to make sure the entity ID, if present, is properly set to the value seen in the
+	// metadata.
+	currentEntityID := sp.GetEntityID()
+	sp.UnsetEntityID()
+
 	if err := sp.CheckAndSetDefaults(); err != nil {
 		return trace.Wrap(err)
 	}
+
+	// If the entity ID was not set, this check is unnecessary.
+	if currentEntityID != "" && currentEntityID != sp.GetEntityID() {
+		return trace.BadParameter("entity ID is set to %s but the entity ID extracted from the provider is %s", currentEntityID, sp.GetEntityID())
+	}
+
+	// Make sure no other service provider has the same entity ID.
```

Comment:
Is this a lint, or an actual invariant that we must maintain for security reasons?

Reply:
This is a lint.

Reply:
I think I'll have the key be the entity ID and try to derive it during creation from the command line. Thoughts on this?

One thing is that the entity ID is not valid as a `name` in Teleport as it's a URI.

Reply:
So this doesn't really apply any more, and the discussion has been hashed out on Slack a good bit -- the key has remained the same and I've stripped down the extraction code to only really apply during the Create call.